functions/snippets/src/main/java/com/example/functions/OcrProcessImage.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.functions;
+
+// [START functions_ocr_process]
+
+import com.google.cloud.functions.BackgroundFunction;
+import com.google.cloud.functions.Context;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.cloud.translate.v3.DetectLanguageRequest;
+import com.google.cloud.translate.v3.DetectLanguageResponse;
+import com.google.cloud.translate.v3.LocationName;
+import com.google.cloud.translate.v3.TranslationServiceClient;
+import com.google.cloud.vision.v1.AnnotateImageRequest;
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.Feature;
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.cloud.vision.v1.ImageSource;
+import com.google.gson.Gson;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.logging.Logger;
+
+public class OcrProcessImage implements BackgroundFunction<GcsEvent> {
+  // TODO<developer> set these environment variables
+  private static final String PROJECT_ID = System.getenv("GCP_PROJECT");
+  private static final String TRANSLATE_TOPIC_NAME = System.getenv("TRANSLATE_TOPIC");
+  private static final String[] TO_LANGS = System.getenv("TO_LANG").split(",");
+
+  private static final Logger LOGGER = Logger.getLogger(OcrProcessImage.class.getName());
+  private static final String LOCATION_NAME = LocationName.of(PROJECT_ID, "global").toString();
+  private static Publisher publisher;
+  private static final Gson gson = new Gson();
+
+  public OcrProcessImage() throws IOException {
+    publisher = Publisher.newBuilder(
+        ProjectTopicName.of(PROJECT_ID, TRANSLATE_TOPIC_NAME)).build();
+  }
+
+  @Override
+  public void accept(GcsEvent gcsEvent, Context context) throws RuntimeException {
```

Comment:
This method seems a bit tiny and pointless as a teaching example - it really only has a single significant line.  I wonder if you might want to do this a bit differently?

Reply:
I was [copying the Node example](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/master/functions/ocr/app/index.js#L113-L136).

(Though I agree this might be worth refactoring - `detectText` is a borderline "God method", so maybe we should move the Pub/Sub publishing code to this method instead?)

Reply:
Let's keep things simple and only ask the user to follow one method.